### PR TITLE
SONARJAVA-1628 Uses LUB when infering vararg types

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/resolve/Types.java [hunk 1]
```diff
@@ -133,6 +135,22 @@ public int compare(Type type, Type t1) {
         return type.name().compareTo(t1.name());
       }
     });
+    Collections.sort(minimalCandidates, new Comparator<Type>() {
+      // Place classes before interfaces
+      @Override
+      public int compare(Type type, Type t1) {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/resolve/Types.java]: This should be merge with the other comparator to have only one pass on the sorting.

File: java-frontend/src/main/java/org/sonar/java/resolve/TypeSubstitutionSolver.java [hunk 2]
```diff
@@ -240,6 +250,22 @@ private TypeSubstitution inferTypeSubstitution(JavaSymbol.MethodJavaSymbol metho
     return currentSubstitution;
   }
 
+  private static JavaType leastUpperBound(List<JavaType> remainingArgTypes) {
+    return (JavaType) Types.leastUpperBound(typeListToTypeWithWrapperSet(remainingArgTypes));
+  }
+
+  private static Set<Type> typeListToTypeWithWrapperSet(List<JavaType> remainingArgTypes) {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeSubstitutionSolver.java]: mapToBoxedSet would be a better name IMO.

File: java-frontend/src/main/java/org/sonar/java/resolve/TypeSubstitutionSolver.java [hunk 3]
```diff
@@ -240,6 +250,22 @@ private TypeSubstitution inferTypeSubstitution(JavaSymbol.MethodJavaSymbol metho
...
+    Set<Type> results = new HashSet<>();
+    for (Type type : remainingArgTypes) {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeSubstitutionSolver.java]: (Very) minor optimisation but you should construct the set and iterate on it,  replacing primitives, that would save some operations.

File: java-frontend/src/main/java/org/sonar/java/resolve/TypeSubstitutionSolver.java [hunk 4]
```diff
@@ -200,16 +205,21 @@ private TypeSubstitution inferTypeSubstitution(JavaSymbol.MethodJavaSymbol metho
...
-    JavaType argType) {
+  private TypeSubstitution inferTypeSubstitution(JavaSymbol.MethodJavaSymbol method, TypeSubstitution currentSubstitution, JavaType formalType, JavaType argType,
+    boolean isLastParam, List<JavaType> remainingArgTypes) {
     if (formalType.isTagged(JavaType.TYPEVAR)) {
       completeSubstitution(currentSubstitution, formalType, argType);
     } else if (formalType.isArray()) {
       JavaType formalElementType = ((JavaType.ArrayJavaType) formalType).elementType;
       if (argType.isArray()) {
-        completeSubstitution(currentSubstitution, formalElementType, ((JavaType.ArrayJavaType) argType).elementType);
+        TypeSubstitution newSubstitution = inferTypeSubstitution(method, currentSubstitution, formalElementType,
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeSubstitutionSolver.java]: this if and the else if can be somehow merged with a null check on the arg type computed (element type of array OR lub or remaining args)

File: java-frontend/src/main/java/org/sonar/java/resolve/TypeSubstitutionSolver.java [hunk 5]
```diff
@@ -200,16 +205,21 @@ private TypeSubstitution inferTypeSubstitution(JavaSymbol.MethodJavaSymbol metho
     return substitution;
   }
 
-  private TypeSubstitution inferTypeSubstitution(JavaSymbol.MethodJavaSymbol method, TypeSubstitution currentSubstitution, boolean isLastParam, JavaType formalType,
-    JavaType argType) {
+  private TypeSubstitution inferTypeSubstitution(JavaSymbol.MethodJavaSymbol method, TypeSubstitution currentSubstitution, JavaType formalType, JavaType argType,
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeSubstitutionSolver.java]: MAJOR Reduce the number of returns of this method 7, down to the maximum allowed 5. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1142)

**Comment by SonarTech** [comment]
#### Analysis summary
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: java-frontend/src/main/java/org/sonar/java/resolve/TypeSubstitutionSolver.java [hunk 6]
```diff
@@ -188,9 +195,10 @@ private TypeSubstitution inferTypeSubstitution(JavaSymbol.MethodJavaSymbol metho
     for (int i = 0; i < Math.min(formals.size(), argTypes.size()); i++) {
       JavaType formalType = formals.get(i);
       JavaType argType = argTypes.get(i);
-      boolean isLastParam = i == formals.size() - 1;
+      boolean checkVariadicType = method.isVarArgs() && i == (formals.size() - 1);
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/resolve/TypeSubstitutionSolver.java]: `variableArity`

**Pull request merged**